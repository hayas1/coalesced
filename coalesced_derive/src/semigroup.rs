use proc_macro2::TokenStream;
use syn::DeriveInput;

use crate::{constant::ConstantExt, semigroup::attr::ContainerAttr};

mod attr;
mod implementor;

pub fn gen_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;

    todo!()
}
