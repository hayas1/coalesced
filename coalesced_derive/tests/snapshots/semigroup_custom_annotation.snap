---
source: coalesced_derive/src/semigroup.rs
assertion_line: 77
expression: formatted
---
#[automatically_derived]
impl ::coalesced::Semigroup for NamedStruct {
    fn semigroup_op(base: Self, other: Self) -> Self {
        Self {
            foo: <coalesced::op::annotation::replace::Replaced<
                _,
            > as ::coalesced::op::Construction<_>>::lift_op(base.foo, other.foo),
            bar: <coalesced::op::annotation::replace::Replaced<
                _,
            > as ::coalesced::op::Construction<_>>::lift_op(base.bar, other.bar),
            baz: <coalesced::op::annotation::replace::Replaced<
                _,
            > as ::coalesced::op::Construction<_>>::lift_op(base.baz, other.baz),
        }
    }
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash)]
pub struct NamedStructAnnotation<X> {
    foo: X,
    bar: X,
    baz: X,
}
#[automatically_derived]
impl<X: Clone> ::coalesced::AnnotatedSemigroup<NamedStructAnnotation<X>>
for NamedStruct {
    fn annotated_op(
        base: ::coalesced::Annotated<Self, NamedStructAnnotation<X>>,
        other: ::coalesced::Annotated<Self, NamedStructAnnotation<X>>,
    ) -> ::coalesced::Annotated<Self, NamedStructAnnotation<X>> {
        let (base_value, base_annotation) = base.into_parts();
        let (other_value, other_annotation) = other.into_parts();
        let (foo_value, foo_annotation) = ::coalesced::AnnotatedSemigroup::annotated_op(
                ::coalesced::Annotated::new(base_value.foo, base_annotation.foo)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
                ::coalesced::Annotated::new(other_value.foo, other_annotation.foo)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
            )
            .into_parts();
        let (bar_value, bar_annotation) = ::coalesced::AnnotatedSemigroup::annotated_op(
                ::coalesced::Annotated::new(base_value.bar, base_annotation.bar)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
                ::coalesced::Annotated::new(other_value.bar, other_annotation.bar)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
            )
            .into_parts();
        let (baz_value, baz_annotation) = ::coalesced::AnnotatedSemigroup::annotated_op(
                ::coalesced::Annotated::new(base_value.baz, base_annotation.baz)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
                ::coalesced::Annotated::new(other_value.baz, other_annotation.baz)
                    .map(coalesced::op::annotation::replace::Replaced::<_>::from),
            )
            .into_parts();
        ::coalesced::Annotated::new(
            NamedStruct {
                foo: <coalesced::op::annotation::replace::Replaced<
                    _,
                > as ::coalesced::op::Construction<_>>::into_inner(foo_value),
                bar: <coalesced::op::annotation::replace::Replaced<
                    _,
                > as ::coalesced::op::Construction<_>>::into_inner(bar_value),
                baz: <coalesced::op::annotation::replace::Replaced<
                    _,
                > as ::coalesced::op::Construction<_>>::into_inner(baz_value),
            },
            NamedStructAnnotation {
                foo: foo_annotation,
                bar: bar_annotation,
                baz: baz_annotation,
            },
        )
    }
}
#[automatically_derived]
impl<X: Clone> ::coalesced::Annotate<NamedStructAnnotation<X>> for NamedStruct {
    type Annotation = X;
    fn annotated(
        self,
        annotation: Self::Annotation,
    ) -> ::coalesced::Annotated<Self, NamedStructAnnotation<X>> {
        ::coalesced::Annotated::new(
            self,
            NamedStructAnnotation {
                foo: annotation.clone(),
                bar: annotation.clone(),
                baz: annotation.clone(),
            },
        )
    }
}
